criu/cr-dump.c

```diff
@@ -1003,6 +1034,228 @@ static int dump_task_signals(pid_t pid, struct pstree_item *item)
+
+	rseqe->has_rseq_cs_pointer = true;
+	rseqe->rseq_cs_pointer = rseq.rseq_cs;
+
+	/* we won't save rseq_cs to the image (only pointer),
+	 * so let's combine flags from both struct rseq and struct rseq_cs
+	 * (kernel does the same when interpreting RSEQ_CS_FLAG_*)
+	 */
+	rseq_cs->flags |= rseq.flags;
+
+	/* save rseq entry to the image */
+	*rseqep = rseqe;
+
+	return 0;
+
+err:
+	xfree(rseqe);
+	return -1;
+}
+
+static int dump_task_rseq(pid_t pid, struct pstree_item *item)
+{
+	int i;
+	struct rseq_cs *thread_rseq_cs;
+
+	/* if rseq() syscall isn't supported then nothing to dump */
+	if (!kdat.has_rseq)
+		return 0;
+
+	thread_rseq_cs = xzalloc(sizeof(*thread_rseq_cs) * item->nr_threads);
+	if (!thread_rseq_cs)
+		return -1;
+
+	dmpi(item)->thread_rseq_cs = thread_rseq_cs;
+
+	for (i = 0; i < item->nr_threads; i++) {
+		if (dump_thread_rseq(item, i))
+			goto free_rseq;
+	}
+
+	return 0;
+
+free_rseq:
+	xfree(thread_rseq_cs);
+	dmpi(item)->thread_rseq_cs = NULL;
+	return -1;
+}
+
+static bool task_in_rseq(struct rseq_cs *rseq_cs, uint64_t addr)
+{
+	return addr >= rseq_cs->start_ip && addr < rseq_cs->start_ip + rseq_cs->post_commit_offset;
+}
+
+static int fixup_thread_rseq(struct pstree_item *item, int i)
+{
+	CoreEntry *core = item->core[i];
+	struct rseq_cs *rseq_cs = &dmpi(item)->thread_rseq_cs[i];
+	pid_t tid = item->threads[i].real;
+
+	/* equivalent to (struct rseq)->rseq_cs is NULL */
+	if (!rseq_cs->start_ip)
+		return 0;
+
+	pr_debug(
+		"fixup_thread_rseq for %d: rseq_cs start_ip = %llx abort_ip = %llx post_commit_offset = %llx flags = %x version = %x; IP = %lx\n",
+		tid, rseq_cs->start_ip, rseq_cs->abort_ip, rseq_cs->post_commit_offset, rseq_cs->flags,
+		rseq_cs->version, (unsigned long)TI_IP(core));
+
+	if (rseq_cs->version != 0) {
+		pr_err("unsupported RSEQ ABI version = %d\n", rseq_cs->version);
+		return -1;
+	}
+
+	if (task_in_rseq(rseq_cs, TI_IP(core))) {
+		struct pid *tid = &item->threads[i];
+
+		/*
+		 * We need to fixup task instruction pointer from
+		 * the original one (which lays inside rseq critical section)
+		 * to rseq abort handler address. But we need to look on rseq_cs->flags
+		 * (please refer to struct rseq -> flags field description).
+		 * Naive idea of flags support may be like... let's change instruction pointer (IP)
+		 * to rseq_cs->abort_ip if !(rseq_cs->flags & RSEQ_CS_FLAG_NO_RESTART_ON_SIGNAL).
+		 * But unfortunately, it doesn't work properly, because the kernel does
+		 * clean up of rseq_cs field in the struct rseq (modifies userspace memory).
+		 * So, we need to preserve original value of (struct rseq)->rseq_cs field in the
+		 * image and restore it's value before releasing threads (see restore_rseq_cs()).
+		 *
+		 * It's worth to mention that we need to fixup IP in CoreEntry
+		 * (used when full dump/restore is performed) and also in
+		 * the parasite regs storage (used if --leave-running option is used,
+		 * or if dump error occurred and process execution is resumed).
+		 */
```

Comment:
>By doing the IP fixup we mimic kernel behavior on preemption (or actually on signal delivery).

Yes, we did IP fixup "by hands" because otherwise, CRIU will preserve an original IP and after the restoration, the process will continue execution from the same place where it was inside the rseq critical section.

>Why do we want to restore rseq_cs? Kernel would set it to NULL, I think we should too. Otherwise there may be a subtle change in behavior introduced by c/r.

The reason is that the kernel wouldn't set `(struct rseq)->rseq_cs` field to `NULL` if the rseq CS has `RSEQ_CS_FLAG_NO_RESTART_ON_SIGNAL` flag set in the normal scenario.

The problem here is that the kernel sets this pointer to `NULL` even when `RSEQ_CS_FLAG_NO_RESTART_ON_SIGNAL` is set when the process leaves CS. In our case that happens during the dump process once CRIU injects and runs the `parasite` from the victim process context. That's why we need to checkpoint and restore this pointer explicitly.

But as I can see I've forgotten to add the check about `RSEQ_CS_FLAG_NO_RESTART_ON_SIGNAL`:
```
	if (read_rseq_cs(tid, &rseqc, rseq_cs, &rseq))
		goto err;

	rseqe->has_rseq_cs_pointer = true;
	rseqe->rseq_cs_pointer = rseq.rseq_cs;

	/* we won't save rseq_cs to the image (only pointer),
	 * so let's combine flags from both struct rseq and struct rseq_cs
	 * (kernel does the same when interpreting RSEQ_CS_FLAG_*)
	 */
	rseq_cs->flags |= rseq.flags;
```

It means that I'm dumping `rseq_cs` pointer in all cases (not only if `RSEQ_CS_FLAG_NO_RESTART_ON_SIGNAL` is set). That does not cause any problem just because, anyway kernel will `NULL`ify this pointer if `RSEQ_CS_FLAG_NO_RESTART_ON_SIGNAL` isn't set. But anyway that's not tidy. Thanks a lot for pointing this out! I've fixed that.